ttnn/cpp/ttnn/device_operation.hpp

```diff
@@ -0,0 +1,319 @@
+    if (not cache_hit) {
+        ZoneScopedN("Program Cache Miss");
+        auto program_factory = operation_t::select_program_factory(operation_attributes, tensor_args);
+
+        auto& program = std::visit(
+            [&program_cache,
+             &program_hash,
+             &operation_attributes,
+             &tensor_args,
+             &tensor_return_value,
+             program_factory_index = program_factory.index()](auto&& program_factory) -> auto& {
+                using program_factory_t = std::decay_t<decltype(program_factory)>;
+                using cached_program_t =
+                    decltype(program_factory_t::create(operation_attributes, tensor_args, tensor_return_value));
+                program_cache.insert(
+                    program_hash,
+                    CachedProgramFactory{
+                        program_factory_t::create(operation_attributes, tensor_args, tensor_return_value),
+                        program_factory_index});
+                auto& cached_program_factory = program_cache.template get<CachedProgramFactory>(program_hash);
+                auto& cached_program = cached_program_factory.cached_program.template get<cached_program_t>();
+                return cached_program.program;
+            },
+            program_factory);
+        return program;
+    } else {
+        ZoneScopedN("Program Cache Hit");
+        auto& cached_program_factory = program_cache.template get<CachedProgramFactory>(program_hash);
+        auto program_factory_index = cached_program_factory.program_factory_index;
+
+        using program_factory_variant_t =
+            decltype(operation_t::select_program_factory(operation_attributes, tensor_args));
+        auto program_factory = map_index_to_variant(program_factory_index, program_factory_variant_t{});
+
+        auto& program = std::visit(
+            [&cached_program_factory, &operation_attributes, &tensor_args, &tensor_return_value](
+                auto&& program_factory) -> auto& {
+                using program_factory_t = std::decay_t<decltype(program_factory)>;
+
+                using cached_program_t =
+                    decltype(program_factory_t::create(operation_attributes, tensor_args, tensor_return_value));
+                auto& cached_program = cached_program_factory.cached_program.template get<cached_program_t>();
+
+                program_factory_t::override_runtime_arguments(
+                    cached_program, operation_attributes, tensor_args, tensor_return_value);
+
+                return cached_program.program;
+            },
+            program_factory);
+        return program;
+    }
+}
+
+template <typename operation_t>
+    requires DeviceOperationConcept<operation_t>
+typename operation_t::tensor_return_value_t run(
+    uint8_t cq_id,
+    const typename operation_t::operation_attributes_t& operation_attributes,
+    const typename operation_t::tensor_args_t& tensor_args) {
+    ZoneScopedN("TT_DNN_DEVICE_OP");
```

Comment:
This might look nicer...

template <DeviceOperationConcept operation_t>
auto run(
    uint8_t cq_id,
    const typename operation_t::operation_attributes_t& operation_attributes,
    const typename operation_t::tensor_args_t& tensor_args
) -> typename operation_t::tensor_return_value_t {